src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -0,0 +1,746 @@
+import org.kohsuke.stapler.StaplerOverridable;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.export.Exported;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+/**
+ * A general-purpose {@link ItemGroup}.
+ * Base for {@link Folder} and {@link ComputedFolder}.
+ * @since FIXME
+ */
+@SuppressWarnings({"unchecked", "rawtypes"}) // mistakes in various places
+public abstract class AbstractFolder<I extends TopLevelItem> extends AbstractItem implements TopLevelItem, ItemGroup<I>, ViewGroup, StaplerFallback, ModelObjectWithChildren, StaplerOverridable {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractFolder.class.getName());
+
+    private static long loadingTick;
+    private static final AtomicInteger jobTotal = new AtomicInteger();
+    private static final AtomicInteger jobEncountered = new AtomicInteger();
+    private static final AtomicBoolean loadJobTotalRan = new AtomicBoolean();
+    private static final int TICK_INTERVAL = 15000;
+
+    @Initializer(before=InitMilestone.JOB_LOADED, fatal=false)
+    public static void loadJobTotal() {
+        if (!loadJobTotalRan.compareAndSet(false, true)) {
+            return; // TODO why does Jenkins run the initializer many times?!
+        }
+        scan(new File(Jenkins.getActiveInstance().getRootDir(), "jobs"), 0);
+        // TODO reset count after reload config from disk (otherwise goes up to 200% etc.)
+    }
+    private static void scan(File d, int depth) {
+        File[] projects = d.listFiles();
+        if (projects == null) {
+            return;
+        }
+        for (File project : projects) {
+            if (!new File(project, "config.xml").isFile()) {
+                continue;
+            }
+            if (depth > 0) {
+                jobTotal.incrementAndGet();
+            }
+            File jobs = new File(project, "jobs"); // cf. getJobsDir
+            if (jobs.isDirectory()) {
+                scan(jobs, depth + 1);
+            }
+        }
+    }
+
+    /** Child items, keyed by {@link Item#getName}. */
+    protected transient Map<String,I> items = new CopyOnWriteMap.Tree<String,I>(CaseInsensitiveComparator.INSTANCE);
+
+    private DescribableList<AbstractFolderProperty<?>,AbstractFolderPropertyDescriptor> properties;
+
+    /**
+     * {@link View}s.
+     */
+    private /*almost final*/ CopyOnWriteArrayList<View> views;
+
+    /**
+     * Currently active Views tab bar.
+     */
+    private volatile ViewsTabBar viewsTabBar;
+
+    /**
+     * Name of the primary view.
+     */
+    private volatile String primaryView;
+
+    private transient /*almost final*/ ViewGroupMixIn viewGroupMixIn;
+
+    private DescribableList<FolderHealthMetric,FolderHealthMetricDescriptor> healthMetrics;
+
+    private FolderIcon icon;
+
+    /** Subclasses should also call {@link #init}. */
+    protected AbstractFolder(ItemGroup parent, String name) {
+        super(parent, name);
```

Comment:
Yeah there is no good way to handle overridable methods to be called from Java constructors. My IDE at least will warn you if you seem to have gotten it wrong.